### PR TITLE
show dollar/euro equivalent

### LINKED ISSUE #199: Show dollar/euro equivalent for amounts

### PR DESCRIPTION
**Pull request by mirecmrozek**
closes #199
getting conversion rates from [CoinGecko](https://www.coingecko.com)

### CODE REVIEW

File: app/frontend/helpers/getConversionRates.js [hunk 1]
```diff
@@ -0,0 +1,10 @@
+const request = require('../wallet/helpers/request')
+
+const getConversionRates = async function getConversionRates() {
```
Comment (refi93) [inline app/frontend/helpers/getConversionRates.js]: I would add caching so each refresh of balance would not trigger an API fetch. I would create a private fetchConversionRates() method that performs the actual fetch and expose getConversionRates() which would then cache the result for let's say 5 minutes. You can inspire by how getAddressInfo and fetchAddressInfo are made in blockchain-explorer.js

File: app/frontend/actions.js [hunk 2]
```diff
@@ -233,13 +236,15 @@ module.exports = ({setState, getState}) => {
     const balance = await wallet.getBalance()
     const ownAddressesWithMeta = await wallet.getOwnAddressesWithMeta()
     const transactionHistory = await wallet.getHistory()
+    const conversionRates = await getConversionRates()
```
Comment (refi93) [inline app/frontend/actions.js]: This is probably a bit dangerous since you don't catch possible exceptions if fetching of the rates fails. I would wrap this into a try-catch block and if fetching fails, I would set the rates to null. If the rates are null, they would not simply display, since it is not critical.

Comment (ppershing) [reply app/frontend/actions.js]: Yep. We even might want to fetch them in parallel so that slow request doesn't block the main wallet

File: app/frontend/components/common/balance.js [hunk 3]
```diff
@@ -13,7 +13,15 @@ const truncatePrintAda = (amount, maxDigits) => {
   return `${Math.trunc(amountString / 1000)}K`
 }
 
-const Balance = ({balance, reloadWalletInfo}) =>
+const printConversionRates = (amount, conversionRates, maxDigits) =>
+  isNaN(Number(amount))
+    ? ''
+    : maxDigits
+      ? `(${truncatePrintAda(printAda(conversionRates.usd * amount), 12)} $,
```
Comment (ppershing) [inline app/frontend/components/common/balance.js]: I am pretty sure `truncatePrintAda(printAda(` combo is not what you want for formatting of fiat currencies. maybe create a separate `formatFiat` function that takes in as an agrument number of decimals?

File: app/frontend/helpers/getConversionRates.js [hunk 4]
```diff
@@ -0,0 +1,10 @@
...
+  const conversionRates = await request(
+    'https://api.coingecko.com/api/v3/coins/cardano?tickers=false&community_data=false&developer_data=false&sparkline=false'
```
[the remark this answers is not in the document]
Comment (refi93) [reply app/frontend/helpers/getConversionRates.js]: perhaps this endpoint is more suitable: https://api.coingecko.com/api/v3/coins/markets?vs_currency=usd&ids=cardano and https://api.coingecko.com/api/v3/coins/markets?vs_currency=eur&ids=cardano

[the remark this answers is not in the document]
Comment (mirecmrozek) [reply app/frontend/helpers/getConversionRates.js]: Thank you for pointing it out!
I found a better Api endpoint: https://min-api.cryptocompare.com/data/price?fsym=USD&tsyms=USD,EUR
Only 29B.

[the remark this answers is not in the document]
Comment (refi93) [reply app/frontend/helpers/getConversionRates.js]: The cryptocompare api looks nice, but @ppershing raised a good point about the terms of service: https://www.cryptocompare.com/api/ - in the case of cryptocompare, if I understand right, we should put a reference to it on our site, which is nevertheless good to do to be transparent about what data we use. I see two suitable locations for the link - either next to the balance or in the footer of the page.

File: app/frontend/components/pages/txHistory/transactionHistory.js [hunk 5]
```diff
@@ -62,6 +66,30 @@ const TransactionHistory = ({transactionHistory}) =>
                 {class: 'align-right'},
                 transaction.effect < 0 ? h('pre', undefined, printAda(transaction.fee)) : ''
               ),
+              h(
```
Comment (refi93) [inline app/frontend/components/pages/txHistory/transactionHistory.js]: adding those fields to the transaction history makes it too cluttered imho. I would rather keep the original structure and have the EUR/USD amounts as titles/tooltips of the the amounts in ADA triggered on hover and on click, so it works properly on mobiles, too. There is already a tooltip element for that in "common" folder

File: app/frontend/helpers/getConversionRates.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (mirecmrozek) [reply app/frontend/helpers/getConversionRates.js]: Right, going to add it to the Footer.
I think "Conversion rates from (link)" would be better as it is shorter and just as clear. What do you think?

[the remark this answers is not in the document]
Comment (refi93) [reply app/frontend/helpers/getConversionRates.js]: looks good to me

**Comment by mirecmrozek** [comment]
This should solve most comments.
I added `padding-top: 1.5em` to `transaction-history-wrapper`, because tooltips on the first line of table were somehow behind the heading. 
Should i still add parallelism to the fetching of conversion rates, now that the request is much smaller?
Also, how do i pass the ci test? It says `Cannot read property 'name' of null` but i don't access `name` anywhere in this PR.

**Comment by ppershing** [comment]
I would keep paralellism, mainly because even if the external service is down we want to keep running (without conversion rates)

**Comment by mirecmrozek** [comment]
Added parallelism to fetching conversion rates.

**Comment by refi93** [comment]
I would round the displayed amounts of USD/EUR to two decimals. Displaying it rounded to 6 decimals seems unnatural

File: app/frontend/actions.js [hunk 6]
```diff
@@ -233,14 +245,26 @@ module.exports = ({setState, getState}) => {
...
     const transactionHistory = await wallet.getHistory()
+    const conversionRates = getConversionRates(state)
 
     // timeout setting loading state, so that loading shows even if everything was cached
     setTimeout(() => setState({loading: false}), 500)
     setState({
       balance,
       ownAddressesWithMeta,
       transactionHistory,
+      conversionRates,
```
Comment (refi93) [inline app/frontend/actions.js]: this is a promise if I understand right, I'm not sure if you want to have a promise there. I think it would be safer to not set the conversionRates there at all and you could move the "const conversionRates = getConversionRates(state)" inside of the try-catch block below then.

I see one particular upside in the approach I suggest - when fetching fails, it does not reset your conversion rates to null when they were set previously. It probably makes sense to keep an older rate rather than delete it completely.

File: app/frontend/actions.js [hunk 7]
```diff
@@ -233,14 +245,26 @@ module.exports = ({setState, getState}) => {
...
     })
+    try {
+      setState({
+        conversionRates: await conversionRates,
```
Comment (refi93) [inline app/frontend/actions.js]: you are doing an  await on an object, not on a promise. I suggest rewriting it like this:

`
let conversionRates = null
try {
    conversionRates = await getConversionRates(state)
} catch (e) {
    debugLog(`Could not fetch conversion rates:  ${e}`)
}
setState({
   balance,
   state,
   etc...
`
